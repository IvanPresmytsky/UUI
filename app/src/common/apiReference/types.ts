import { TTypeRef, TTypeSummary } from './sharedTypes';

export type TDocsGenExportedType = Autogenerated_TDocsGenExportedTypeRef;
export type TDocsGenTypeSummary = Record<TTypeRef, TTypeSummary>;
