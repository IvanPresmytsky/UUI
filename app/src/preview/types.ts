import { TComponentPreview } from '@epam/uui-docs';

export type TPreviewContentParams = {
    theme: string;
    isSkin: boolean;
    componentId: string | undefined;
    previewId: string | undefined | TComponentPreview<unknown>;
};

export type TPredefinedPreviewRefItem = { groupId?: string; id: string, link: string };
export type TPreviewRef = { link: string, error: string | undefined, predefinedPreviewRefs: TPredefinedPreviewRefItem[] };
