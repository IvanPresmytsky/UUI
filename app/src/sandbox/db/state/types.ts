import { Person, PersonGroup } from '@epam/uui-docs';

export type PersonTableRecord = Person | PersonGroup;
