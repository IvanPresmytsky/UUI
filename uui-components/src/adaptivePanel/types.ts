import { IHasCX, IHasRawProps } from '@epam/uui-core';

export type AdaptiveItemProps<T = unknown> = T & {
    /** Render callback of the item. It renders items inside the panel and measures their width.
     * Pay attention that we can't measure margins. If you need to have margins, please make a wrapper and add margins inside
     * */
    render: (item: AdaptiveItemProps<T>, hiddenItems?: AdaptiveItemProps<T>[], displayedItems?: AdaptiveItemProps<T>[]) => any;
    /** Item collapsed priority. An item with lower priority will be hidden first, and so on. If several items have the same priority, all of them will be hidden; */
    priority: number;
    /** If true, this item will be shown when some other items was hidden; for example, you can use it to render dropdowns with hidden items.
     * You can provide more than one collapsedContainer item, but will be shown only those which has minimal priority, but this priority can't be less than the last hidden itemâ€™s priority. */
    collapsedContainer?: boolean;
    /** Unique ID of item */
    id: string;
};

export interface AdaptivePanelProps extends IHasCX, IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    /** Array of items to be rendered in AdaptivePanel */
    items: AdaptiveItemProps[];
}
