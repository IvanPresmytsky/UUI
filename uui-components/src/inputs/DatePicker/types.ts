import * as React from 'react';
import { Dayjs } from 'dayjs';
import {
    IHasCX, IHasRawProps, RangeDatePickerPresets, IHasForwardedRef, DayProps,
} from '@epam/uui-core';

export type ViewType = 'DAY_SELECTION' | 'MONTH_SELECTION' | 'YEAR_SELECTION';

export interface DatePickerBodyBaseOptions extends IHasCX, IHasRawProps<React.HTMLAttributes<HTMLDivElement>>, IHasForwardedRef<HTMLDivElement> {
    filter?(day: Dayjs): boolean;
    presets?: RangeDatePickerPresets;
    renderDay?: (renderProps: DayProps) => React.ReactElement<Element>;
}

export interface DatePickerBodyBaseProps<TSelection> extends DatePickerBodyBaseOptions {
    value: PickerBodyValue<TSelection>;
    onValueChange: (value: PickerBodyValue<TSelection>) => void;
}
export interface PickerBodyValue<TSelection> {
    selectedDate: TSelection | null;
    month: Dayjs;
    view: ViewType;
}

export interface RangePickerBodyValue<TSelection> extends PickerBodyValue<TSelection> {
    inFocus: 'from' | 'to';
}
