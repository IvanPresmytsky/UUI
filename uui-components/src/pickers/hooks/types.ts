import {
    DataSourceState, IEditable, IHasCaption, IHasRawProps, IModal, PickerBaseOptions, PickerBaseProps,
    PickerFooterProps, PickerInputBaseProps, SortingOption,
} from '@epam/uui-core';
import { Dispatch, SetStateAction } from 'react';

export type UsePickerInputProps<TItem, TId, TProps> = PickerInputBaseProps<TItem, TId> & TProps & {
    toggleModalOpening?(opened: boolean): void;
    shouldShowBody?(): boolean;
};

export interface UsePickerInputStateProps extends UsePickerStateProps {}

export interface PickerInputState extends PickerState {
    opened: boolean;
    setOpened: Dispatch<SetStateAction<boolean>>;
    isSearchChanged: boolean;
    setIsSearchChanged: Dispatch<SetStateAction<boolean>>;
}

export interface UsePickerStateProps {
    dataSourceState?: Partial<DataSourceState>;
}

export interface PickerState {
    dataSourceState: DataSourceState;
    setDataSourceState: Dispatch<SetStateAction<DataSourceState>>;
    showSelected: boolean;
    setShowSelected: Dispatch<SetStateAction<boolean>>;
}

export type PickerListBaseProps<TItem, TId> = Exclude<PickerBaseProps<TItem, TId>, 'cascadeSelection'> & {
    /**
     * Number of default items to show initially, when nothing is selected.
     * @default 10
     */
    maxDefaultItems?: number;

    /**
     * Maximum total number of items to show, including selected
     * @default 50
     */
    maxTotalItems?: number;

    /**
     * Ids of items to show first.
     * If not specified, top props.maxDefaultItems will be shown according to the DataSource sorting settings (default is 10)
     */
    defaultIds?: TId[];

    /** If provided, top picks will be automatically adjusted based on last user selection, and stored as user setting under provided key */
    settingsKey?: string;

    sortBy?(item: TItem, sorting: SortingOption): string;
};

export type UsePickerListProps<TItem, TId, TProps> = PickerListBaseProps<TItem, TId> & TProps & {};

export interface PickerModalOptions<TItem, TId> extends IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    /** Render callback for custom filters block. This block will be rendered befor the items list.
     If omitted, nothing will be rendered.
     */
    renderFilter?(editableFilter: IEditable<any>): React.ReactNode;
    /** Render callback for modal footer.
     If omitted, default footer will be rendered.
     */
    renderFooter?: (props: PickerFooterProps<TItem, TId> & Partial<IModal<any>>) => React.ReactNode;
    /**
     * If true, prevent modal window closing by click outside modal
     * @default false
     */
    disallowClickOutside?: boolean;
}

export type PickerModalScalarProps<TId, TItem> =
    | ({ selectionMode: 'single'; valueType: 'id'; initialValue: TId } & IModal<TId>)
    | ({ selectionMode: 'single'; valueType: 'entity'; initialValue: TItem } & IModal<TItem>);
export type PickerModalArrayProps<TId, TItem> =
    | ({ selectionMode: 'multi'; valueType: 'id'; initialValue: TId[] } & IModal<TId[]>)
    | ({ selectionMode: 'multi'; valueType: 'entity'; initialValue: TItem[] } & IModal<TItem[]>);

export type UsePickerModalProps<TItem, TId> = PickerBaseOptions<TItem, TId> &
IHasCaption &
(PickerModalScalarProps<TId, TItem> | PickerModalArrayProps<TId, TItem>) &
PickerModalOptions<TItem, TId>;
