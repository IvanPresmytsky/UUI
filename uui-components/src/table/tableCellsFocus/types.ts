import { DataTableCellProps } from '@epam/uui-core';
import type { DataTableFocusManager } from './DataTableFocusManager';

export type RowsRegistry<TId> = Map<TId | string, Array<CellInfo>>;

export interface DataTableFocusManagerProps {
    enableMovementShortcuts?: boolean;
}

export type CellProps = Pick<DataTableCellProps, 'index' | 'key' | 'isReadonly' | 'isDisabled'>;

export interface CellFocusAPI {
    focus: () => void;
}

export interface CellInfo {
    ref: React.RefObject<CellFocusAPI>,
    cellProps: CellProps;
}

export interface RowInfo<TId> {
    id: TId,
    index: number;
}

export type { DataTableFocusManager };
