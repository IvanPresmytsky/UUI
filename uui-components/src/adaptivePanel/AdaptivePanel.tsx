import React, { useEffect, useRef, useState } from 'react';
import css from './AdaptivePanel.module.scss';
import { FlexRow } from '../layout/flexItems';
import { measureAdaptiveItems } from './measureItemsUtils';
import { useLayoutEffectSafeForSsr } from '@epam/uui-core';
import { AdaptivePanelProps } from './types';
import cx from 'classnames';

export function AdaptivePanel(props: AdaptivePanelProps) {
    const [itemsWidth, setItemsWidth] = useState<Record<string, number>>();
    const [isChanged, setIsChanged] = useState(false);
    const wrapperRef = useRef<HTMLDivElement>(null);
    const displayedRowRef = useRef<HTMLDivElement>(null);

    const getItemsWidth = () => {
        if (!displayedRowRef.current) {
            return;
        }
        const children = Array.from(displayedRowRef.current.children);

        if (!children.length) return;
        const calculatedItemsWidth: Record<string, number> = {};
        children.forEach((child, index) => {
            calculatedItemsWidth[props.items[index].id] = child.getBoundingClientRect().width;
        });

        return calculatedItemsWidth;
    };

    useLayoutEffectSafeForSsr(() => {
        if (isChanged || !itemsWidth) {
            const newItemsWidth = getItemsWidth();
            setItemsWidth(newItemsWidth);
            setIsChanged(false);
        }
    });

    useEffect(() => {
        const resizeObserver = new ResizeObserver((entries) =>
            requestAnimationFrame(() => {
                if (!Array.isArray(entries) || !entries.length) return;
                setIsChanged(true);
            }));

        resizeObserver.observe(displayedRowRef.current);
        resizeObserver.observe(wrapperRef.current);
        return () => {
            resizeObserver.disconnect();
        };
    }, []);

    const renderItems = () => {
        if (isChanged || !itemsWidth) {
            return props.items.map((i) => i.render(i, [], props.items));
        }
        const wrapperWidth = wrapperRef?.current ? wrapperRef.current.getBoundingClientRect().width : 0;

        const measuredItems = measureAdaptiveItems(props.items, wrapperWidth, itemsWidth);
        return measuredItems.displayed.map((i) => i.render(i, measuredItems.hidden, measuredItems.displayed));
    };

    return (
        <div { ...props.rawProps } className={ cx(props.cx, css.mainWrapper) } ref={ wrapperRef }>
            <FlexRow ref={ displayedRowRef }>{ renderItems() }</FlexRow>
        </div>
    );
}
