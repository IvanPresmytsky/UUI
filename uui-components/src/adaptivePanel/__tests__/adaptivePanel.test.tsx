import { AdaptivePanel } from '../AdaptivePanel';
import { AdaptiveItemProps } from '../types';
import {
    renderSnapshotWithContextAsync, renderWithContextAsync, mockAdaptivePanelLayout, screen,
} from '@epam/uui-test-utils';
import React from 'react';

type TestItemType = AdaptiveItemProps<{ data?: { label: string } }>;

function getNItems(baseId: string, n: number, priority: number = 1): TestItemType[] {
    const genItem = (params: Pick<TestItemType, 'id' | 'priority'>): TestItemType => {
        return {
            id: params.id,
            priority: params.priority,
            collapsedContainer: false,
            data: { label: `Item ${params.id}` },
            render: (item) => <div data-testid="adaptive-item" data-label={ item.data.label } key={ item.id } />,
        };
    };
    const items: TestItemType[] = [];
    for (let i = 0; i < n; i++) {
        items.push(genItem({ id: `${baseId}-${i}`, priority }));
    }
    return items;
}

async function setupAdaptivePanel({ width, itemWidth }: { width: number; itemWidth: number }) {
    const genContainerItem = ({ id, priority }: Pick<TestItemType, 'id' | 'priority'>): TestItemType => {
        return {
            id,
            priority,
            collapsedContainer: true,
            data: { label: `Collapsed Container Item ${id}` },
            render: (item, hiddenItems) => (
                <div key={ id } data-testid="adaptive-item-cc" data-label={ item.data.label } data-hiddenitems={ hiddenItems.map((i) => i.id).join(',') } />
            ),
        };
    };
    const items: TestItemType[] = [
        ...getNItems('p1', 3, 1), ...getNItems('p2', 3, 3), ...getNItems('p3', 3, 5), genContainerItem({ id: 'wide-screen', priority: 2 }), genContainerItem({ id: 'medium-screen', priority: 4 }), genContainerItem({ id: 'small-screen', priority: 6 }),
    ];
    mockAdaptivePanelLayout({ width, itemWidth });
    const result = await renderWithContextAsync(<AdaptivePanel items={ items } rawProps={ { 'data-testid': 'adaptive-panel' } } />);
    const visibleItems = screen
        .queryAllByTestId('adaptive-item')
        .map((i) => i.getAttribute('data-label'))
        .join(',');
    const containerItem = screen.queryByTestId('adaptive-item-cc');
    const hiddenItems = containerItem.getAttribute('data-hiddenitems');
    return {
        result,
        visibleItems,
        hiddenItems,
        containerItem,
    };
}

describe('AdaptivePanel', () => {
    it('should render no items provided', async () => {
        const tree = await renderSnapshotWithContextAsync(<AdaptivePanel items={ [] } />);
        expect(tree).toMatchSnapshot();
    });

    it('should render some items', async () => {
        const tree = await renderSnapshotWithContextAsync(<AdaptivePanel items={ getNItems('snapshot', 5) } />);
        expect(tree).toMatchSnapshot();
    });

    it('should properly collapse on WIDE screen', async () => {
        const res = await setupAdaptivePanel({ width: 140, itemWidth: 20 });
        expect(res.visibleItems).toEqual('Item p2-0,Item p2-1,Item p2-2,Item p3-0,Item p3-1,Item p3-2');
        expect(res.containerItem).toBeInTheDocument();
        expect(res.containerItem.getAttribute('data-label')).toEqual('Collapsed Container Item wide-screen');
        expect(res.hiddenItems).toEqual('p1-0,p1-1,p1-2');
    });

    it('should properly collapse on MEDIUM screen', async () => {
        const res = await setupAdaptivePanel({ width: 100, itemWidth: 20 });
        expect(res.visibleItems).toEqual('Item p3-0,Item p3-1,Item p3-2');
        expect(res.containerItem).toBeInTheDocument();
        expect(res.containerItem.getAttribute('data-label')).toEqual('Collapsed Container Item medium-screen');
        expect(res.hiddenItems).toEqual('p1-0,p1-1,p1-2,p2-0,p2-1,p2-2');
    });

    it('should properly collapse on SMALL screen', async () => {
        const res = await setupAdaptivePanel({ width: 30, itemWidth: 20 });
        expect(res.visibleItems).toEqual('');
        expect(res.containerItem).toBeInTheDocument();
        expect(res.containerItem.getAttribute('data-label')).toEqual('Collapsed Container Item small-screen');
        expect(res.hiddenItems).toEqual('p1-0,p1-1,p1-2,p2-0,p2-1,p2-2,p3-0,p3-1,p3-2');
    });
});
