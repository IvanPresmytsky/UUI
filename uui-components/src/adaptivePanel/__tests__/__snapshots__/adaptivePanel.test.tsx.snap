// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdaptivePanel should render no items provided 1`] = `
<div
  className="mainWrapper"
>
  <div
    className="itemsWrapper"
    style={
      Object {
        "columnGap": undefined,
      }
    }
  />
</div>
`;

exports[`AdaptivePanel should render some items 1`] = `
<div
  className="mainWrapper"
>
  <div
    className="itemsWrapper"
    style={
      Object {
        "columnGap": undefined,
      }
    }
  >
    <div
      data-label="Item snapshot-0"
      data-testid="adaptive-item"
    />
    <div
      data-label="Item snapshot-1"
      data-testid="adaptive-item"
    />
    <div
      data-label="Item snapshot-2"
      data-testid="adaptive-item"
    />
    <div
      data-label="Item snapshot-3"
      data-testid="adaptive-item"
    />
    <div
      data-label="Item snapshot-4"
      data-testid="adaptive-item"
    />
  </div>
</div>
`;

exports[`AdaptivePanel should render some items with itemsGap 1`] = `
<div
  className="mainWrapper"
>
  <div
    className="itemsWrapper"
    style={
      Object {
        "columnGap": 6,
      }
    }
  >
    <div
      data-label="Item snapshot-0"
      data-testid="adaptive-item"
    />
    <div
      data-label="Item snapshot-1"
      data-testid="adaptive-item"
    />
    <div
      data-label="Item snapshot-2"
      data-testid="adaptive-item"
    />
    <div
      data-label="Item snapshot-3"
      data-testid="adaptive-item"
    />
    <div
      data-label="Item snapshot-4"
      data-testid="adaptive-item"
    />
  </div>
</div>
`;
